go-controller/pkg/ovn/namespace.go

```diff
@@ -135,112 +134,137 @@ func (oc *Controller) multicastUpdateNamespace(ns *kapi.Namespace) {
 	} else {
 		for _, pod := range existingPods {
 			if pod.Status.PodIP != "" {
 				portName := podLogicalPortName(pod)
-				oc.namespaceAddressSet[ns.Name][pod.Status.PodIP] = portName
+				nsInfo.addressSet[pod.Status.PodIP] = portName
+				addresses = append(addresses, pod.Status.PodIP)
 			}
 		}
 	}
 
-	addresses := make([]string, 0)
-	for address := range oc.namespaceAddressSet[ns.Name] {
-		addresses = append(addresses, address)
-	}
-
 	// Create an address_set for the namespace.  All the pods' IP address
 	// in the namespace will be added to the address_set
 	createAddressSet(ns.Name, hashedAddressSet(ns.Name), addresses)
 
-	oc.namespacePolicies[ns.Name] = make(map[string]*namespacePolicy)
-	oc.multicastUpdateNamespace(ns)
+	oc.multicastUpdateNamespace(ns, nsInfo)
 }
 
 func (oc *Controller) updateNamespace(old, newer *kapi.Namespace) {
-	klog.V(5).Infof("Updating namespace: old %s new %s", old.Name, newer.Name)
+	klog.V(5).Infof("Updating namespace: %s", old.Name)
 
-	// A big fat lock per namespace to prevent race conditions
-	// with namespace resources like address sets and deny acls.
-	oc.namespaceMutex[newer.Name].Lock()
-	defer oc.namespaceMutex[newer.Name].Unlock()
+	nsInfo := oc.getNamespaceLocked(old.Name)
+	if nsInfo == nil {
+		klog.Warningf("Update event for unknown namespace %q", old.Name)
+		return
+	}
+	defer nsInfo.Unlock()
 
-	oc.multicastUpdateNamespace(newer)
+	oc.multicastUpdateNamespace(newer, nsInfo)
 }
 
 func (oc *Controller) deleteNamespace(ns *kapi.Namespace) {
 	klog.V(5).Infof("Deleting namespace: %s", ns.Name)
-	oc.namespaceMutexMutex.Lock()
-	defer oc.namespaceMutexMutex.Unlock()
 
-	mutex, ok := oc.namespaceMutex[ns.Name]
-	if !ok {
+	nsInfo := oc.deleteNamespaceLocked(ns.Name)
+	if nsInfo == nil {
 		return
 	}
-	mutex.Lock()
-	defer mutex.Unlock()
+	defer nsInfo.Unlock()
 
 	deleteAddressSet(hashedAddressSet(ns.Name))
-	oc.multicastDeleteNamespace(ns)
-	delete(oc.namespacePolicies, ns.Name)
-	delete(oc.namespaceAddressSet, ns.Name)
-	delete(oc.namespaceMutex, ns.Name)
+	oc.multicastDeleteNamespace(ns, nsInfo)
 }
 
-// getNamespaceLock grabs the lock for a particular namespace. If the
-// namespace does not exist, returns nil. Otherwise, returns the held lock.
-func (oc *Controller) getNamespaceLock(ns string) *sync.Mutex {
-	// lock the list of namespaces, get the mutex
-	oc.namespaceMutexMutex.Lock()
-	mutex, ok := oc.namespaceMutex[ns]
-	oc.namespaceMutexMutex.Unlock()
-	if !ok {
+// getNamespaceLocked locks namespacesMutex, looks up ns, and (if found), returns it with
+// its mutex locked. If ns is not known, nil will be returned
+func (oc *Controller) getNamespaceLocked(ns string) *namespaceInfo {
+	oc.namespacesMutex.Lock()
+	nsInfo := oc.namespaces[ns]
+	oc.namespacesMutex.Unlock()
+
+	if nsInfo == nil {
+		return nil
+	}
+	nsInfo.Lock()
+
+	oc.namespacesMutex.Lock()
+	defer oc.namespacesMutex.Unlock()
+
+	// We may have blocked on getting nsInfo's mutex, and the namespace may have been
```

Comment:
cant you get around this by locking nsinfo before you release namespacesMutex lock? Just defer the unlock and i think you can simplify the function like this:

```
func (oc *Controller) getNamespaceLocked(ns string) *namespaceInfo {
	oc.namespacesMutex.Lock()
        defer oc.namespacesMutex.Unlock()
        if nsInfo, ok := oc.namespaces[ns]; ok {
             nsInfo.Lock()
             return nsInfo
        }
        return nil
```

Reply:
If we do that though, and someone else is doing something "slow" with `nsInfo`, then we'll be holding on to `namespacesMutex` until they're done, and then no other threads can call `getNamespaceLocked` for *any* namespace.

Are there actual examples of such "slow" operations that would cause problems like this in practice? I dunno, but the old `getNamespaceLock` code did it this way, so I kept that behavior...

Reply:
yeah didnt think about nsInfo.Lock taking a while, that makes sense

Reply:
Yeah, I added that logic to the old code. Some of the code makes `nbctl` calls while holding the namespace's lock, and we don't want that to block all other pending work.

Reply:
especially because both the namespace and pod handlers can claim the namespace lock.